Sources/SwiftSyntax/Identifier.swift

```diff
@@ -26,41 +29,67 @@ public struct Identifier: Equatable, Hashable, Sendable {
   public init?(_ token: TokenSyntax) {
     switch token.tokenKind {
     case .identifier, .keyword(.self), .keyword(.Self):
-      self.raw = RawIdentifier(token.tokenView)
+      self.raw = RawIdentifier(token.tokenView.rawText)
       self.arena = token.raw.arenaReference
-
-      self.dollarIdentifierStr = nil
     case .dollarIdentifier(let dollarIdentifierStr):
-      self.raw = RawIdentifier(token.tokenView)
       self.arena = token.raw.arenaReference
 
-      self.dollarIdentifierStr = dollarIdentifierStr
+      if Self.isPaddedDollarIdentifier(dollarIdentfierStr: dollarIdentifierStr),
+        let newDollarIdentifierNumber = Int(dollarIdentifierStr.dropFirst())
+      {
+        let newDollarIdentifierStr = "$\(newDollarIdentifierNumber)"
+        let sanitizedDollarIdentifierSyntaxText = token.raw.arenaReference.intern(newDollarIdentifierStr)
+
+        self.raw = RawIdentifier(sanitizedDollarIdentifierSyntaxText)
+      } else {
+        self.raw = RawIdentifier(token.tokenView.rawText)
+      }
     default:
       return nil
     }
   }
 
-  public init(_ staticString: StaticString) {
-    self.raw = RawIdentifier(staticString)
-    self.arena = nil
-
-    let name = String(syntaxText: raw.name)
+  /// Create a new `Identifier` from given `staticString`.
+  ///
+  /// - Precondition: `staticString` is a canonical identifier i.e. doesn't
+  ///   use backticks and is not a dollar identifier with leading zeros.
+  public init(staticString: StaticString) {
```

Comment:
What’s the reason why you added the `staticString` argument label?

Reply:
I noticed this used to produce a compilation error
```swift
let identifier = Identifier("smth") // Error: Ambiguous use of 'init(_:)'
```
Interestingly, these two worked absolutely fine
```swift
let x: StaticString = "smth"
let identifier1 = Identifier(x)
let identifier2: Identifier = Identifier("smth")
```
It seemed to me a bit odd overloading the initializer worked like that in this case. Is there a reason where a string literal could be used with the other `init?(_ token: TokenSyntax)`?

This behavior doesn't happen with an explicit argument label so that's why I changed that. Do you think there could be a better way to fix this?

Reply:
Oh, I see. `TokenSyntax` is expressible by string literal as well, that’s why it’s ambiguous.

I’m not a fan of the `staticString` label. How about we use the argument label to convey that the string needs to be canonical? That way a user will get the overload via `TokenSyntax` that canonicalizes the name by default and they need to opt into this faster but more restrictive overload.

Ie. change it to `init(canonicalName:)` or something of the sort.

Reply:
Oh that makes sense why it didn't work with string literals, thanks for explanation. I also like the idea of `init(canonicalName:)`. I changed it.